Modernize util/strencodings and util/string: `string_view` and `optional`

maflcko opened the pull request:
Make use of `std::string_view` and `std::optional` in the util/{strencodings, string} files.

This avoids many temporary string/vector objects being created, while making the interface easier to read. Changes include:
* Make all input arguments in functions in util/strencodings and util/string take `std::string_view` instead of `std::string`.
* Add `RemovePrefixView` and `TrimStringView` which also *return* `std::string_view` objects (the corresponding `RemovePrefix` and `TrimString` keep returning an `std::string`, as that's needed in many call sites still).
* Stop returning `std::string` from `DecodeBase32` and `DecodeBase64`, but return vectors. Base32/64 are fundamentally algorithms for encoding bytes as strings; returning `std::string` from those (especially doing it conditionally based on the input arguments/types) is just bizarre.
* Stop taking a `bool* pf_invalid` output argument pointer in `DecodeBase32` and `DecodeBase64`; return an `std::optional` instead.
* Make `DecodeBase32` and `DecodeBase64` more efficient by doing the conversion from characters to integer symbols on-the-fly rather than through a temporary vector.

sipa committed:
Make ParseHex use string_view

sipa committed:
Make IsHex use string_view

sipa committed:
Make IsHexNumber use string_view

sipa committed:
Make SanitizeString use string_view

sipa committed:
Generalize ConvertBits to permit transforming the input

sipa committed:
Make DecodeBase{32,64} always return vector, not string

Base32/base64 are mechanisms for encoding binary data. That they'd
decode to a string is just bizarre. The fact that they'd do that
based on the type of input arguments even more so.

sipa committed:
Make DecodeBase{32,64} return optional instead of taking bool*

sipa committed:
Make DecodeBase{32,64} take string_view arguments

sipa committed:
Use std::string_view throughout util strencodings/string

sipa committed:
Reject incorrect base64 in HTTP auth

In addition, to make sure that no call site ignores the invalid
decoding status, make the pf_invalid argument mandatory.

MarcoFalke commented:
Disclaimer: The description and most commits are stolen from https://github.com/bitcoin/bitcoin/pull/24764#issuecomment-1110922692

MarcoFalke commented:
Previous review: https://github.com/bitcoin/bitcoin/pull/24764#pullrequestreview-937752519

re-ACK fa7078d84fc2858a466bc1a85404f821df682538 only change is rebase and adding a scripted-diff 🍲

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK fa7078d84fc2858a466bc1a85404f821df682538 only change is rebase and adding a scripted-diff 🍲
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjYUQv/TURxeLbk3gryj4kDPgdoqT5a/1EOIBJ7kAM9bT/jzpxuqM9ZNOGSlIxu
DmmVgwGwMxvxE0fd7WjxOrdYTSTIQfMwlPAJpSR3JCTHvh/xI1ew1Xmjq1iRTgcY
6ma3XMavD9X54VbpH2tlMRj3F/Vhb5MHAQiJUU3v3vImst8BL+HfwB3fsXkYlsPe
F5auWj58clyrtGRed5HnL4UDKLmLg95S/9jSUFh/6sIgxw7QNWbXhnNxtBDjuOLw
EDpZd9NmhDBqmq6dx9wPQMEoYMDjeICXLciMzoUU6WI+Y7UuK5kdHrF+Y6/Jogqc
Q9CLGkBOxUOACc+i115Q+SKykiibphLYvIrbSxv8etjbB4XFQNvJKL96UCUP9pa5
h4vy/k9V2hspW9KRoazJevv90rb3m7sKG7yqRr4IurWcfTWW+oPv3Wff9uXbd4oQ
AfQg2ql9wCHcC+3T/1o9GDJOkSf1xCDUCD/wReekhr8JuLiJhVvg9bUkUuPwOUa0
Ddfn99lR
=dVEV
-----END PGP SIGNATURE-----
```


</details>

Commit:
scripted-diff: Rename ValidAsCString to ContainsNoNUL

-BEGIN VERIFY SCRIPT-
 sed -i 's,ValidAsCString,ContainsNoNUL,g' $(git grep -l ValidAsCString)
-END VERIFY SCRIPT-

laanwj commented:
Code review ACK  fa7078d84fc2858a466bc1a85404f821df682538

sipa commented:
utACK fa7078d84fc2858a466bc1a85404f821df682538 (as far as the commit that isn't mine goes)
